Allow users to load into any Python code from GPI (#1293)
cocotb/cocotb: documentation/source/python_entry_point.rst

Python Entry Point
==================

> By default, cocotb starts the :class:`RegressionManager` as the testing framework for cocotb. However, there are many reason why a user might want to supplant the whole or part of the existing Python framework with an alternative. To this end, we have defined :envvar:`PYTHON_ENTRY`, to allow the user to enter into an alternative testing framework.

Reviewer: ```suggestion
By default, cocotb starts the :class:`RegressionManager` as the testing framework for cocotb. However, there are many reasons why a user might want to supplant the whole or part of the existing Python framework with an alternative. To this end, we have defined :envvar:`PYTHON_ENTRY`, to allow the user to enter into an alternative testing framework.
```

> By default, cocotb starts the :class:`RegressionManager` as the testing framework for cocotb. However, there are many reason why a user might want to supplant the whole or part of the existing Python framework with an alternative. To this end, we have defined :envvar:`PYTHON_ENTRY`, to allow the user to enter into an alternative testing framework.

Reviewer 2: ```suggestion
By default, cocotb starts the :class:`~cocotb.regression.RegressionManager` as the testing framework for cocotb. 
However, there are many reasons why a user might want to supplant the whole or part of the existing Python framework with an alternative.
To this end, we have defined :envvar:`COCOTB_PYTHON_ENTRY`, to allow the user to enter into an alternative testing framework.
```

.. versionadded:: 2.0

How It Works
------------

> After GPI implementations (VPI, VHPI, FLI, etc.) are registered, the Python implementation to the GPI (henceforth, PyGPI) starts a Python interpreter, loads the ``cocotb`` module, gets references to necessary helper functions from that module, and then calls into the ``_initialise_testbench`` function in that module to start the regression manager. The :envvar:`PYTHON_ENTRY` variable allows the user to specify a module and function other than ``cocotb._initialise_testbench`` that contains the alternative testing framework.

Reviewer 2: We have slowly americanized the spelling also of functions etc. as we touched them, so maybe better ``_initialize_testbench``.

Author: This is a name of an existing function. Should the function be renamed in concert?

Reviewer 2: Existing functions should probably stay the same, I didn't realize that.

How To Use It
-------------

.. currentmodule:: cocotb

PyGPI expects the following functions to be implemented in the entry module. The names are required.

.. autofunction:: _sim_event

.. autofunction:: _log_from_c

.. autofunction:: _filter_from_c

> Finally, there is the entry function itself. ``_initialise_testbench`` is being used as an example, the entry function is not required to be named the same. A single entry module can even contain multiple entry functions.

Reviewer 2: Maybe it's just me, but could someone read this last sentence as "A single-entry module can even contain multiple entry functions." and be confused for a second?
Would "One entry module can even contain multiple entry functions." circumvent that and be still correct?

Reviewer: I agree it could be ambiguous. What about "**The** entry module can even contain multiple entry functions."

Reviewer 2: Indeed, that sounds better.

Custom Entry Point
==================

By default, cocotb starts the regression manager as the testing framework for cocotb.
However, there are many reasons why a user might want to replace the whole or a part of the existing Python framework with an alternative.
To support this, the environment variable :envvar:`COCOTB_ENTRY` can be used.

> This allow the user to specify the location of an alternative testing framework to run at initialization.

Reviewer 2: ```suggestion
This allows the user to specify the location of an alternative testing framework to run at initialization.
```

Python Entry Point
==================

> By default, cocotb starts the :class:`~cocotb.regression.RegressionManager` as the testing framework for cocotb.

Reviewer 2: This isn't hyperlinked because ``RegressionManager`` is not in the Library Reference chapter - would it make sense to add it?

Author: I don't know if any of the regression manager was meant to be touched by the user and documenting it makes changing it even harder. Should I remove the ref instead?

However, there are many reasons why a user might want to replace the whole or a part of the existing Python framework with an alternative.

> To support this, the environment variable :envvar:`PYTHON_ENTRY` was created.

Reviewer 2: ```suggestion
To support this, the environment variable :envvar:`PYTHON_ENTRY` can be used.
```
The "was created" makes sense now in this PR, but in the future, the variable is just there, and referring to the creation may sound... quirky? (Not a native speaker, so my feeling may be wrong.)

This allow the user to specify the location of an alternative testing framework to run at initialization.

.. versionadded:: 2.0

How It Works
------------

After GPI implementations (VPI, VHPI, FLI, etc.) are registered,
the Python interface to the GPI (henceforth, PyGPI) does the following:
 - starts the Python interpreter
 - loads the ``cocotb`` module
 - gets references to necessary helper functions from the module
 - calls the ``_initialize_testbench`` function in the module
 - ``_initialize_testbench`` starts the regression manager.
The :envvar:`PYTHON_ENTRY` variable allows the user to specify a module and function other than ``cocotb._initialize_testbench`` that contains the alternative testing framework.
See the documentation for :envvar:`PYTHON_ENTRY` for details on the entry point specification.

Entry Point Interface Requirements
----------------------------------

.. currentmodule:: cocotb

PyGPI expects the following functions to be implemented in the entry module. The names are required.

.. autofunction:: _sim_event

.. autofunction:: _log_from_c

.. autofunction:: _filter_from_c

> 

Reviewer 3: I know I've already expressed this elsewhere, for but reviewers who haven't seen my comments - I don't think we want to do this. These functions are designed specifically to connect to the Python `logging` infrastructure, and to so me it doesn't make a whole lot of sense to allow users to intercept them. If users want to configure logging, then they can do it through the builtin `logging` module.

The only reason they're defined in python at all is it makes things a little easier to write them. Idealistically, I'd consider them part of the same unit as `gpi_embed.c`. I wouldn't be opposed to moving them to `cocotb/_gpi_embed.py` to indicate this.

Author: This feature was never really intended to allow users to override the behavior, just allow them to run without invoking the cocotb Python module. Whether they are in `cocotb.log` or `cocotb._gpi_embed`, `cocotb.__init__` is still run. 

You are right though, after the refactoring these functions are not tied to cocotb's logger in any way and probably could be moved into C. Writing C API code sucks, is it possible we could define the functions in Python in a C string and compile them into Python function objects from `gpi_embed`?

Reviewer 3: > just allow them to run without invoking the cocotb Python module

One option would be to create a `_simulator_gpi_embed.py` file, and use that.

> `cocotb.__init__` is still run.

perhaps the issue then is that `cocotb.__init__` being run is stateful. Presumably if it ran but did nothing other than define functions, you would have no issues with it?

Author: Presumably if it ran but did nothing other than define functions, you would have no issues with it?
It's better, and what I want to eventually do, but what's C's should be C's in this case. This can be pushed to another day. I'll revert the commit and continue pointing to `cocotb.log`.

Reviewer 3: > but what's C's should be C's in this case

What I'm modelling this after is `numpy.core._internal`, which contains random python functions that are needed by the C code, and are far too much effort to write in C. I don't think we should be afraid of pushing details to python as long as those details remain private.

> I'll revert the commit and continue pointing to cocotb.log. 

Since this is a draft, feel free to keep iterating on the rest before doing that.

Author: I would like to refactor the module, but I don't want it too balloon too much. Large PRs tend to not get pulled in.

Author: I have decided to move as much processing from `gpi_embed.c` into a `_gpi_embed.py`
